File: 3. Longest Substring Without Repeating Characters.py
class Solution(object):
    def lengthOfLongestSubstring(self, s):
        #trying to use 'pointers' and hashmaps for this one
        hashmap = {}
        left = 0
        right = 0
        maxL = 0
        for right in range(len(s)):
            if s[right] in hashmap:
                left = max(left, hashmap[s[right]] + 1)
            hashmap[s[right]] = right
            maxL = max(maxL, right - left + 1)
        return maxL

#really enjoyed this one, the sorta 'sliding box' defined by the left and right 'pointers' are really different than anything ive done before
#15ms beats 85.85%, but as with 2. the exact execution time is quite random due to python. I do believe this is pretty close to the fastest way possible besides small things

File: 3. Longest Substring Without Repeating Characters.go
func lengthOfLongestSubstring(s string) int {
    hashmap := make(map[byte]int)
    var left int
    var right int
    var maxL int
    for right=0; right < len(s); right++ {
        if _, found := hashmap[s[right]]; found {
            left = max(left, hashmap[s[right]] + 1)
        }
        hashmap[s[right]] = right
        maxL = max(maxL, right - left + 1)
    }
    return maxL
}
// i just completely copied my python solution but in go to try to practice more with go hashmaps

// 0ms beats 100%
